substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/service/GcNotifier.java

```diff
@@ -0,0 +1,162 @@
+    GcNotifier() {
+        // Pre-allocate the queue so we can use it later from allocation free code.
+        requests = new GcNotificationRequest[QUEUE_SIZE];
+        for (int i = 0; i < QUEUE_SIZE; i++) {
+            requests[i] = new GcNotificationRequest(GenScavengeMemoryPoolMXBeans.singleton().getMXBeans().length);
+        }
+        currentRequest = new GcNotificationRequest(GenScavengeMemoryPoolMXBeans.singleton().getMXBeans().length);
+        rear = 0;
+        front = 0;
+    }
+
+    /** Called during GC. */
+    void beforeCollection(long startTime) {
+        assert VMOperation.isInProgressAtSafepoint();
+
+        AbstractMemoryPoolMXBean[] beans = GenScavengeMemoryPoolMXBeans.singleton().getMXBeans();
+        for (int i = 0; i < beans.length; i++) {
+            /*
+             * Always add the old generation pool even if we don't end up using it (since we don't
+             * know what type of collection will happen yet)
+             */
+            requests[rear].setPoolBefore(i, beans[i].getUsedBytes().rawValue(), beans[i].getCommittedBytes().rawValue(), beans[i].getName());
+        }
+        requests[rear].startTime = startTime;
+    }
+
+    /** Called during GC. */
+    void afterCollection(boolean isIncremental, GCCause cause, long epoch, long endTime) {
+        assert VMOperation.isInProgressAtSafepoint();
+
+        AbstractMemoryPoolMXBean[] beans = GenScavengeMemoryPoolMXBeans.singleton().getMXBeans();
+        for (int i = 0; i < beans.length; i++) {
+            if (isIncremental && beans[i].getName().equals(OLD_GEN_SPACE)) {
+                continue;
+            }
+            requests[rear].setPoolAfter(i, beans[i].getUsedBytes().rawValue(), beans[i].getCommittedBytes().rawValue(), beans[i].getName());
+        }
+        requests[rear].endTime = endTime;
+        requests[rear].isIncremental = isIncremental;
+        requests[rear].cause = cause;
+        requests[rear].epoch = epoch;
+        requests[rear].timestamp = System.currentTimeMillis();
+
+        rear = incremented(rear);
+
+        // If rear is now lapping front, then increment front to prevent overlap. Accept data loss.
+        if (front == rear) {
+            front = incremented(front);
+        }
+    }
+
+    @Uninterruptible(reason = Uninterruptible.CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+    private static int incremented(int value) {
+        return (value + 1) % QUEUE_SIZE;
+    }
+
+    /**
+     * Called by the service thread. Sends a notification if any are queued. If none are queued,
+     * then return immediately.
+     */
+    public void sendNotification() {
+        assert !VMOperation.isInProgressAtSafepoint();
+
+        /*
+         * We don't need to drain all requests since signals queue at the semaphore level. The rest
+         * of the requests will be handled in subsequent calls to this method by the sevice thread.
+         */
+        if (!updateCurrentRequest()) {
```

Comment:
Please adapt this logic once you switch to `VMMutex` and `VMCondition`.

Reply:
This is now handled by checking `NotificationThread.hasRequests()` before blocking in the notification thread loop.